src/sundials/sundials_timestepheuristics.c

```diff
@@ -0,0 +1,376 @@
+ * Programmer(s): Daniel R. Reynolds @ SMU
+ * -----------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------
+ * This is the implementation file for a generic
+ * SUNTimestepHeuristics package. It contains the implementation of
+ * the SUNTimestepHeuristics operations listed in
+ * sundials_timestepheuristics.h
+ * -----------------------------------------------------------------*/
+
+#include <sundials/sundials_timestepheuristics.h>
+#include <sundials/sundials_nvector.h>
+
+/* -----------------------------------------------------------------
+ * Create a new empty SUNTimestepHeuristics object
+ * ----------------------------------------------------------------- */
+
+SUNTimestepHeuristics SUNTimestepHeuristics_NewEmpty(SUNContext sunctx)
+{
+  SUNTimestepHeuristics H;
+  SUNTimestepHeuristics_Ops ops;
+
+  /* a context is required */
+  if (sunctx == NULL) return(NULL);
+
+  /* create heuristics object */
+  H = NULL;
+  H = (SUNTimestepHeuristics) malloc(sizeof *H);
+  if (H == NULL) return(NULL);
+
+  /* create matrix ops structure */
+  ops = NULL;
+  ops = (SUNTimestepHeuristics_Ops) malloc(sizeof *ops);
+  if (ops == NULL) { free(H); return(NULL); }
+
+  /* initialize operations to NULL */
+  ops->getid              = NULL;
+  ops->destroy            = NULL;
+  ops->constrainstep      = NULL;
+  ops->etestfail          = NULL;
+  ops->convfail           = NULL;
+  ops->boundreduction     = NULL;
+  ops->boundfirststep     = NULL;
+  ops->reset              = NULL;
+  ops->update             = NULL;
+  ops->setdefaults        = NULL;
+  ops->write              = NULL;
+  ops->setmaxstep         = NULL;
+  ops->setminstep         = NULL;
+  ops->setexpstabfn       = NULL;
+  ops->setcflfraction     = NULL;
+  ops->setsafetyfactor    = NULL;
+  ops->setmaxgrowth       = NULL;
+  ops->setminreduction    = NULL;
+  ops->setfixedstepbounds = NULL;
+  ops->setmaxfirstgrowth  = NULL;
+  ops->setmaxefailgrowth  = NULL;
+  ops->setsmallnumefails  = NULL;
+  ops->setmaxcfailgrowth  = NULL;
+  ops->getnumexpsteps     = NULL;
+  ops->getnumaccsteps     = NULL;
+  ops->space              = NULL;
+
+  /* attach ops and initialize content to NULL */
+  H->ops     = ops;
+  H->content = NULL;
+  H->sunctx  = sunctx;
+
+  return(H);
+}
+
+
+/* -----------------------------------------------------------------
+ * Required functions in the 'ops' structure
+ * ----------------------------------------------------------------- */
+
+SUNTimestepHeuristics_ID SUNTimestepHeuristics_GetID(SUNTimestepHeuristics H)
+{
+  if (H == NULL) { return SUN_TIMESTEPHEURISTICS_NULL; }
```

Comment:
Same comment as on the controller.  I dont think we want to make `NULL` a valid "object."